### DISCUSSION TITLE
Ubuntu Installation (Java Error)

### DISCUSSION THREAD

**Discussion by zealbell** [beehive-lab/TornadoVM#218] [Q&A]
Hi, Please I am having volatile issues configuring **TornadoVM** on Ubuntu 22.10. I cloned the 'beehive-lab/TornadoVM' repository using Intellij and then installed asper 

`$ ./scripts/tornadoVMInstaller.sh --graal-jdk-17 --opencl --ptx` (since: AMD RYZEN 9 5950X + RTX 3080)

 I then opened the module `TornadoVM/tornado-examples` as a new project but then the build fails with the message 

```java
java: error: release version 5 not supported

Module tornado-examples SDK 17 does not support source version 1.5. 
Possible solutions:
- Downgrade Project SDK in settings to 1.5 or compatible. Open project settings.
- Upgrade language version in Maven build file to 17. Update pom.xml and reload the project.

```

(Tried possible solutions above, I got NOTHING, so I reverted changes made). Currently my project's (`.../TornadoVM/tornado-examples`) sdk is @ `.../TornadoVM/etc/dependencies/TornadoVM-GraalJDK17`

Also at the pom.xml the truffle dependency just goes red in between the tags ( `<groupId>`, `<artifactId>` and `<version>`) then I noticed jdk-8 in between the `<id>` tags. I am just a bit confused and suspicious  about what 8 has to do with installing graaljdk-17 earlier, could this be the reason my build isn't successful? could it just be 8? (I am still yet to understand how 5 came about earlier)

```xml
 <profiles>
        <profile>
            <id>jdk-8</id>
            <activation>
                <activeByDefault>false</activeByDefault>
            </activation>
            <dependencies>
                <dependency>
                    <groupId>org.graalvm.truffle</groupId>
                    <artifactId>truffle-api</artifactId>
                    <version>${graal.version}</version>
                </dependency>
            </dependencies>
        </profile>
    </profiles>
```

**Comment by jjfumero** [comment]
JDK 8 is not supported. The two JDK versions supported are 11 and 17, and we highly recommend staying with 17 when possible. 

After you install Tornadovm using the installer, you can recompile your projects from the terminal:

```bash
cd <TornadoVM-Root-Directory>
. source.sh
make graal-jdk-11-plus BACKEND=opencl,ptx 
```

**Reply by zealbell** [reply]
Hi @jjfumero , thanks  for the reply. I just want to make a few clarifications.  I am only able to create a maven project  that builds successfully with [this](https://github.com/beehive-lab/TornadoVM#5-how-to-use-tornadovm-in-your-projects) after trying a couple things here and there. I installed tornado again 
(asper `$ ./scripts/tornadoVMInstaller.sh --jdk11 --opencl` after I did `$ ./scripts/tornadoVMInstaller.sh --graal-jdk-17 --opencl --ptx`) . I understand that whenever I open my maven project in Intellij I have to do this initially (at least before hitting the run button):
```bash
$ cd <TornadoVM-Root-Directory>
$ . source.sh
$ make jdk-11 BACKEND=opencl
$ mvn clean install
```
But what if I would like to change my jdk and backend (currently its  `jdk11` and `opencl`). Say I want it to be  `graal-jdk-17` and `opencl` + `ptx`.  (Since this was what I installed tornado earlier with). I tried this:
```bash
$ make graal-jdk-17 BACKEND=opencl,ptx  ##I was hoping this command would also update the source.sh file, since the variables there changed
make: *** No rule to make target 'graal-jdk-17'.  Stop.  ##Got this as result

```
How can I achieve this without needing to re-install/re-download tornado objects again (since some of the files (`graal-jdk-17` objects) from the initial installation are still present at `.../TornadoVM/etc/dependencies`). Thanks

**Reply by jjfumero** [reply]
Not sure I followed. Every time the installer script is executed, the installer will override the `source.sh` directory. All paths and env variables are in that file (`JAVA_HOME`, `TORNADO_SDK`, etc). So if you change your JDK, you need to update your JAVA_HOME by invoking `. source.sh`. 

If you want to have different JDK options, you can have a different `source` file for each JDK.

For example: 

```bash
./scripts/tornadoVMInstaller.sh --jdk11 --opencl
mv source.sh jdk11source.sh 
./scripts/tornadoVMInstaller.sh --jdk17 --opencl
mv source.sh jdk17source.sh 

# Use JDK 11:
. jdk11source.sh 
tornado .... 

# Use JDK 17:
. jdk17source.sh 
tornado .... 
```

Additionally, once you installed TornadoVM with any JDK, you do not need to execute the installer again to add/remove backends. 
You can just recompile as follows (e.g., if you are using Graal):

```bash
 make graal-jdk-11-plus BACKEND=ptx
```

**Reply by zealbell** [reply]
Hi, you followed :). Would it be correct to do this (after creating different sources):

```bash
$ cd <TornadoVM-Root-Directory>
$ . jdk17source.sh
$ make graal-jdk-17 BACKEND=ptx
$ mvn clean install
```

Lastly I want to ask about `make` in essence since you said
> if you are using Graal):
> `make graal-jdk-11-plus BACKEND=ptx`

is it only applicable with graal only asper would doing this:

```bash
$ cd <TornadoVM-Root-Directory>
$ . source.sh
$ make jdk-11 BACKEND=opencl ##make with just jdk-11?
$ mvn clean install
```
be considered illegal, thanks

**Reply by jjfumero** [reply]
> Hi, you followed :). Would it be correct to do this (after creating different sources):
> 
> ```shell
> $ cd <TornadoVM-Root-Directory>
> $ . jdk17source.sh
> $ make graal-jdk-17 BACKEND=ptx
> $ mvn clean install
> ```
> 

Lines 1-3, yes. Line 4 (`mvn clean install`) You don't need this. The `make` process will do it for you. 
Note that the `make` build process is just more convenient for us to easy the configuration process. In reality, it is just a chain of scripts that link OpenCL, SPIRV and PTX configuration plus all paths needed for setup TornadoVM.  

> Lastly I want to ask about `make` in essence since you said
> 
> > if you are using Graal):
> > `make graal-jdk-11-plus BACKEND=ptx`
> 
> is it only applicable with graal only asper would doing this:
> 
> ```shell
> $ cd <TornadoVM-Root-Directory>
> $ . source.sh
> $ make jdk-11 BACKEND=opencl ##make with just jdk-11?
> $ mvn clean install
> ```
> 
> be considered illegal, thanks

What do you mean by illegal? As a rule? This is just a `Makefile` 

For JDK 11 you use the `jdk-11-plus` rule

```bash
make jdk-11-plus BACKEND=opencl  ## Build with JDK >= 11 
```

**Reply by zealbell** [reply]
Thanks alot🔥

**Reply by zealbell** [reply]
Hi again, I just wanted to ask about how I can `make` with `graal-jdk-17`.

```bash
$ make graal-jdk-17 BACKEND=ptx  
make: *** No rule to make target 'graal-jdk-17'.  Stop.  ##I am still getting this

```
So I took a look at the `Makefile` and noticed there’s no rule like that present

**Reply by stratika** [reply]
> Hi again, I just wanted to ask about how I can `make` with `graal-jdk-17`.
> 
> ```shell
> $ make graal-jdk-17 BACKEND=ptx  
> make: *** No rule to make target 'graal-jdk-17'.  Stop.  ##I am still getting this
> ```
> 
> So I took a look at the `Makefile` and noticed there’s no rule like that present

hi @54LiNKeR, you have to use the `graal-jdk-11-plus` rule in Makefile for any JDK version >= `11`.
Try: ``` $ make graal-jdk-11-plus BACKEND=ptx```, it should work.

**Reply by zealbell** [reply]
working… thanks🙏🏾
